hypothesis-python/src/hypothesis/database.py

```diff
@@ -675,6 +677,43 @@ def delete(self, key: bytes, value: bytes) -> None:
         raise RuntimeError(self._read_only_message)
 
 
+class BackgroundWriteDatabase(ExampleDatabase):
+    """A wrapper which defers writes on the given database to a background thread.
+    Reads are still executed synchronously on the calling thread.
+    """
+
+    def __init__(self, db: ExampleDatabase) -> None:
+        self._db = db
+        self._queue: Queue[tuple[str, tuple[bytes, ...]]] = Queue()
+        self._thread = Thread(target=self._worker, daemon=True)
+        self._thread.start()
+
+    def __repr__(self) -> str:
+        return f"BackgroundWriteDatabase({self._db!r})"
+
+    def _worker(self) -> None:
+        while True:
+            method, args = self._queue.get()
+            getattr(self._db, method)(*args)
+            self._queue.task_done()
+
+    def fetch(self, key: bytes) -> Iterable[bytes]:
+        return self._db.fetch(key)
```

Comment:
I think we don't want to deal with retrieving enqeueued-but-not-finalized values because then we have to replicate the semantics of `move` + `delete`, and may start disagreeing with an unsound database which doesn't respect the method contracts. Which wouldn't really be our fault, but still.

I think joining on fetch would be reasonable? (note that the queue is only writes, fetches aren't enqueued). Possibly we want a `join: bool = True` or `timeout: int | None = None` parameter.

Reply:
"join before fetch" cleans up test usages as well, which makes me more confident this is the correct default behavior.